File: sdk/typescript/examples/basic-usage.ts
/**
 * Basic SDK Usage Example
 *
 * Demonstrates connecting to the Aethelred network, querying node info,
 * listing seals and jobs, and fetching blocks.
 *
 * Run with: npx ts-node examples/basic-usage.ts
 */

import {
  AethelredClient,
  Network,
  AethelredError,
  ConnectionError,
} from '../src';

async function main() {
  console.log('='.repeat(60));
  console.log('        Aethelred SDK - Basic Usage Example');
  console.log('='.repeat(60));
  console.log();

  // Create a client targeting the testnet
  const client = new AethelredClient({ network: Network.TESTNET });

  // 1. Health check
  console.log('Checking node health...');
  const healthy = await client.healthCheck();
  console.log(`  Node healthy: ${healthy}`);
  console.log();

  // 2. Node information
  console.log('Fetching node info...');
  try {
    const nodeInfo = await client.getNodeInfo();
    console.log('Node Info:');
    console.log(`  Network : ${nodeInfo.network}`);
    console.log(`  Moniker : ${nodeInfo.moniker}`);
    console.log(`  Version : ${nodeInfo.version}`);
  } catch (err) {
    if (err instanceof ConnectionError) {
      console.log(`  Could not reach node: ${err.message}`);
    } else {
      throw err;
    }
  }
  console.log();

  // 3. Latest block
  console.log('Fetching latest block...');
  try {
    const block = await client.getLatestBlock();
    console.log('Latest Block:');
    console.log(`  Height : ${block.block?.header?.height ?? 'N/A'}`);
    console.log(`  Time   : ${block.block?.header?.time ?? 'N/A'}`);
    console.log(`  Chain  : ${block.block?.header?.chain_id ?? 'N/A'}`);
  } catch (err) {
    if (err instanceof AethelredError) {
      console.log(`  Error fetching block: ${err.message}`);
    } else {
      throw err;
    }
  }
  console.log();

  // 4. Query seals
  console.log('Querying recent seals...');
  try {
    const sealList = await client.seals.list({ limit: 5 });
    console.log(`Found ${sealList.total} total seals`);

    if (sealList.seals.length > 0) {
      console.log('\nRecent Seals:');
      for (const seal of sealList.seals) {
        console.log(`  - ${seal.id}`);
        console.log(`    Status: ${seal.status}`);
        console.log(`    Purpose: ${seal.purpose}`);
        console.log(`    Block: ${seal.blockHeight}`);
      }
    }
  } catch (err) {
    if (err instanceof AethelredError) {
      console.log(`  Error querying seals: ${err.message}`);
    } else {
      throw err;
    }
  }
  console.log();

  // 5. Query compute jobs
  console.log('Querying recent compute jobs...');
  try {
    const jobList = await client.jobs.list({ limit: 5 });
    console.log(`Found ${jobList.total} total jobs`);

    if (jobList.jobs.length > 0) {
      console.log('\nRecent Jobs:');
      for (const job of jobList.jobs) {
        console.log(`  - ${job.id}`);
        console.log(`    Status: ${job.status}`);
        console.log(`    Proof Type: ${job.proofType}`);
      }
    }
  } catch (err) {
    if (err instanceof AethelredError) {
      console.log(`  Error querying jobs: ${err.message}`);
    } else {
      throw err;
    }
  }
  console.log();

  // 6. Network configuration
  console.log('Network Configuration:');
  console.log(`  RPC URL  : ${client.getRpcUrl()}`);
  console.log(`  Chain ID : ${client.getChainId()}`);
  console.log();

  console.log('='.repeat(60));
  console.log('  Basic usage example complete.');
  console.log('='.repeat(60));
}

main().catch(console.error);

File: sdk/go/examples/basic_usage.go
// Basic usage of the Aethelred Go SDK.
//
// Demonstrates client creation, health check, node info, and listing
// jobs and seals from the testnet.
//
// Build: go build -o basic_usage ./examples/basic_usage.go
// Run:   AETHELRED_API_KEY=<key> ./basic_usage
package main

import (
	"context"
	"fmt"
	"log"
	"os"
	"time"

	"github.com/aethelred/sdk-go/client"
)

func main() {
	ctx := context.Background()

	// Create a testnet client with an API key from the environment.
	apiKey := os.Getenv("AETHELRED_API_KEY")
	c, err := client.NewClient(
		client.Testnet,
		client.WithAPIKey(apiKey),
		client.WithTimeout(30*time.Second),
	)
	if err != nil {
		log.Fatalf("Failed to create client: %v", err)
	}

	fmt.Println("RPC URL: ", c.RPCURL())
	fmt.Println("Chain ID:", c.ChainID())

	// Health check -- returns true when the node responds.
	healthy := c.HealthCheck(ctx)
	fmt.Printf("Node healthy: %v\n", healthy)

	// Get node info
	info, err := c.GetNodeInfo(ctx)
	if err != nil {
		log.Fatalf("Failed to get node info: %v", err)
	}
	fmt.Printf("Network:  %s\n", info.Network)
	fmt.Printf("Version:  %s\n", info.Version)
	fmt.Printf("Moniker:  %s\n", info.Moniker)

	// List compute jobs
	fmt.Println("\n--- Compute Jobs ---")
	jobs, err := c.Jobs.List(ctx, nil)
	if err != nil {
		log.Fatalf("Failed to list jobs: %v", err)
	}
	fmt.Printf("Found %d job(s)\n", len(jobs))
	for _, j := range jobs {
		fmt.Printf("  Job %s | Status: %s | Proof: %s\n",
			j.ID, j.Status, j.ProofType)
	}

	// List digital seals
	fmt.Println("\n--- Digital Seals ---")
	seals, err := c.Seals.List(ctx, nil)
	if err != nil {
		log.Fatalf("Failed to list seals: %v", err)
	}
	fmt.Printf("Found %d seal(s)\n", len(seals))
	for _, s := range seals {
		fmt.Printf("  Seal %s | Status: %s | Job: %s\n",
			s.ID, s.Status, s.JobID)
	}
}
